### ISSUE TITLE
LoadClass - ClassNotFountException / NoClassDefFoundError

### ISSUE THREAD

**Issue by rgoncalves-Bib** [classgraph/classgraph#222]
Hi,

When I call loadClass on a class' field that implement an interface from an external JAR, I get a ClassNotFountException / NoClassDefFoundError (depending on how I configure the scanner). Is this a normal behaviour ? I can't get to understand how this comes, as when it's called straight from the class, I don't have this issue. 

Also, is it possible to open the method loadClass(boolean ignoreException) as public. It would prevent users from having a ugly try{...}catch(Exception e)

Thanks for the help !

**Comment by lukehutch** [comment]
No problem, but I should say that I'm expecting classloading issues to be the #1 problem to come up over the long term. It's a very complex and difficult issue, and I have a couple of reasonable solutions built into ClassGraph, but I don't know that they will work all of the time. So thanks for the report anyway.

The `loadClass(boolean)` method is already public... 

https://github.com/classgraph/classgraph/blob/master/src/main/java/io/github/classgraph/ClassInfo.java#L1657

I think it was always public too. Just wondering why you can't call it?

**Comment by rgoncalves-Bib** [comment]
Thank you for your feed back.

The loadClass comes from ClassRefTypeSignature which doesn't the `loadClass(boolean)` signature (even if going deeper into the call hierarchy, it calls the `ClassInfo.loadClass(boolean)` method.

It's not that much of a problem. Just an idea :smile:

**Comment by lukehutch** [comment]
@rgoncalves-Bib Ah, I wondered if this would ever become a problem! There are (I think) two places where a type signature might need to load a class reference, and I didn't have a convenient way to always pass parameters in there. Maybe I should make those safe, and just ignore exceptions by default... Can you please give me a little bit more info about your usecase (e.g. code snippets), so that I can make sure I understand exactly the issue you're raising?

**Comment by lukehutch** [comment]
I just committed what I think are all the missing `loadClass(boolean)` methods, did I get them all? Was the issue that you called `loadClass()` yourself, and it threw an exception? Or did I miss a place in the ClassGraph code that internally called `loadClass()` and threw an exception?

**Comment by rgoncalves-Bib** [comment]
Yes I call call `loadClass()`by myself. I had to do this to be able to load my Enum values when the type was a Enum. 

```java
if (type != null && ((Class) type).isEnum()){
      val = EnumUtils.getEnumList((Class) type);
}
```

**Comment by lukehutch** [comment]
That code example doesn't have a `loadClass()` in it, and in fact you are always dealing with `Class`, not `ClassInfo`, (the class was already loaded).

If you have an enum constant in an `AnnotationParameterValue`, you can call `AnnotationEnumValue#loadClassAndReturnEnumValue(boolean)` to load it:

```
Object value = annotationParameterValue.getValue();
Object enumValue = null;
if (value instanceof AnnotationEnumValue) {
    enumValue = ((AnnotationEnumValue) value)
        .loadClassAndReturnEnumValue(/* IgnoreExceptions = */ true);
}
```

Not sure if this even applies to your situation...

But either way, please check if the commit shown above (4392a91) adds the missing methods you would expect.

**Comment by rgoncalves-Bib** [comment]
Yes, sorry !  The above piece of code is called before. 

```java
if(typeSignature instanceof BaseTypeSignature) {
            BaseTypeSignature signature = (BaseTypeSignature)typeSignature;
            typeStr = signature.getType().getName();
            type = signature.getType();
        }
        else if(typeSignature instanceof ClassRefTypeSignature) {
            typeStr = ((ClassRefTypeSignature) typeSignature).getBaseClassName();
            type = ((ClassRefTypeSignature) typeSignature).loadClass();
        }
```

To give more information about my use case, I go through each classes which have a given annotation. Then go through all fields to retrieve their information (types, annotations + annotations values). 

------

Still, the commit seems ok for where I needed it :)

**Comment by lukehutch** [comment]
Released as 4.0.1. Thanks for pointing out the missing API calls!

**Comment by rgoncalves-Bib** [comment]
Thanks for the information ! And for the quick release (once again). The tool is really great !

**Comment by lukehutch** [comment]
Thanks, glad you like it!

By the way, the code snippet you most recently shared pretty much exactly duplicates the logic in `AnnotationClassRef`, so I assume that you are reading class reference values in annotations?

https://github.com/classgraph/classgraph/blob/4392a919dcdf065bc384148226997b1a3a373e0b/src/main/java/io/github/classgraph/AnnotationClassRef.java#L85

This sets the `type` variable the same way as your code does currently, so there's no need to duplicate it.

As for the `typeStr` string, you should call `AnnotationClassRef#getClassName()`. This gets the fully-qualified name though (including inner classes), whereas your code only gets the base name. I think you actually want the fully-qualified name anyway, in case it's an inner class.

https://github.com/classgraph/classgraph/blob/4392a919dcdf065bc384148226997b1a3a373e0b/src/main/java/io/github/classgraph/AnnotationClassRef.java#L112

If not, should `AnnotationClassRef`

**Comment by lukehutch** [comment]
PS re quick releases, they're probably a pain for some people, because the version number keeps increasing every few days... but I figure that's much better than multiple people getting hit with the same bug :-)

**Comment by rgoncalves-Bib** [comment]
Well, I am working with FieldInfo class, which has no method to load the type/class directly. 
That is why I do : 

```java
        this.field = field; //FieldInfo
        TypeSignature typeSignature = field.getTypeDescriptor();
        String typeStr = null;
        Type type = null;
        isBaseType = false;
        if(typeSignature instanceof BaseTypeSignature) {
            BaseTypeSignature signature = (BaseTypeSignature)typeSignature;
            typeStr = signature.getType().getName();
            type = signature.getType();
            isBaseType = true;
        }
        else if(typeSignature instanceof ClassRefTypeSignature) {
            typeStr = ((ClassRefTypeSignature) typeSignature).getBaseClassName();
            type = ((ClassRefTypeSignature) typeSignature).loadClass();
        }
```

**Comment by rgoncalves-Bib** [comment]
(my code is not perfect as I am still in development discovering mode with the scanner)

**Comment by lukehutch** [comment]
Don't worry, I'm not judging your code, just pointing out cases where the API might already do what you're looking for :-)

I was wrong though, if you're working with fields, not annotation parameter values, then you're doing the right thing, since `TypeSignature` does not have a `.loadClass()` method, or `.getClassName()` or `.getClassInfo()` methods, because one of the subclasses of `TypeSignature`, `TypeVariableSignature`, doesn't necessarily have an associated class (you have to call `TypeVariableSignature#resolve()` to resolve it against its base class to get the type of the type variable).

You do need to check if `typeSignature` is an instance of `ArrayTypeSignature` or `TypeVariableSignature`, since these are valid types for fields too (for arrays or generic-typed fields respectively).
